### ISSUE TITLE
Lifting a block with self-modifying support on

### ISSUE THREAD

**Issue by symike** [angr/angr#552]
Hello,

I currently have a `Project` created with `support_selfmodifying_code=True`. The binary loaded by the `Project` is not even self-modifying. I'm executing with `PathGroup`'s `step()`, and then iterating through the active paths.

When I try to lift a block as such: `project.factory.block(path.addr, backup_state=path.state)`, I observe the following exception:

`simuvex.s_errors.SimEngineError: No bytes in memory for block starting at 0x0.`

So, for some reason it's trying to read from 0x0. However, the execution continues just fine. When executing `project.factory.block(path.addr)`, without setting `backup_state` to the current state, I observe the same issue.

Thanks,
Mike

**Comment by zardus** [comment]
To preface, as of a few days ago, `PathGroup` no longer exists (the equivalent is `SimulationManager`). There is backwards-compatible API, though.

As a secondary preface, @rhelmot is the core individual of the self-modifying code stuff, so I might be spouting some incorrect material here :-)

At any rate, I suspect that the problem isn't in the self modifying code, but rather in angr's environment support, and something that was supposed to initialize a code pointer did not do so. There are a few steps for debugging:

1. look in the trace (path.trace.hardcopy in your version, or state.history.descriptions on master) to see if there are any ReturnUnconstrained procedures. This means that some library function is missing from angr's environment model, and you might have to actually load and execute the library (leading to problem 2).
2. Look in the trace to see if there is an unsupported syscall (if you're bypassing unsupported syscalls).
3. Look at what the code is doing and track it back to see where it _should_ have been initialized, then break there (using SimInspect or `simgr.explore` to get there) to see what's going on.

**Comment by zardus** [comment]
Also, as a sanity check: can you confirm that `path.addr` is 0?

**Comment by symike** [comment]
Thanks @zardus. It turns out that it's indeed not the self-modifying code support that's directly causing this. It turns out that using the `num_inst=1` option during execution is causing it (which is something forced by the self-modifying support).

It seems that it's a different issue (this one can be renamed, or closed and I can open a new one). Since there might still be an issue, I'll answer your questions:


There are no `ReturnUnconstrained` procedures. The program contains calls to `printf`, which in turn should call the `write` syscall. These are both supported, as far as I know. I've also tried with the `auto_load_libs=True` option, but I observe the same behavior.

I'll look into updating to `master`, and start using `SimulationManager`, though I believe the underlying execution has not changed that much (apart from the new interface), as this is a recent revision.

`path.addr` is indeed 0 (so is `path.state.regs.rsi`).

**Comment by symike** [comment]
Any updates on what might cause this? It's happening only for PE files.

**Comment by rhelmot** [comment]
Hi! I genuinely have no idea how to help you here. I have never seen anything like this happen. Especially weird is the part where you say "execution continues just fine", even though path.addr is 0. Execution should definitely not continue just fine in this case.

Can you do anything else to help us help you?

**Comment by symike** [comment]
I've cooked up an example. It actually seems to happen for both PE and ELF files. I've tried to make a tiny script that shows this problem. I've also updated to the latest version, and used the `SimulationManager` interface, to make sure it's not some issue that was already fixed.

I've attached scripts,` angr_test_simgr.py` (using the `SimulationManager` interface) and `angr_test.py` (using the deprecated `PathGroup` interface). I've also attached `print1`, which the smallest binary I could think of that shows this behavior. The binary is the result of compiling with GCC (to a 64-bit ELF) a program that simply calls `printf("%d\n", 0x12345)`.

Running each of these programs with the `print1` binary, as such:

> ` python angr_test.py print1`
> ` python angr_test_simgr.py print1`

results in:

> angr.errors.SimEngineError: No bytes in memory for block starting at 0x0.

Sorry, @ "execution continues just fine", just ignore that.

[test_to_show_error.zip](https://github.com/angr/angr/files/1241682/test_to_show_error.zip)

**Comment by rhelmot** [comment]
Ok! This is incredibly dumb.

So this is specifically the fault of the intersection between fastpath mode and single-stepping. Fastpath mode is an extremely lightweight execution mode designed for use during CFG recovery that aims to do as little as possible. Most relevantly, it doesn't execute ccalls. You can see exactly what "fastpath" means at the bottom of angr/sim_options.py.

Since calculating the eflags at every step is expensive, VEX takes a shortcut and doesn't actually calculate/store the flags, it just stores the last operation that changed the flags along with its operands. When it gets to an instruction that uses the flags, like a branch, it calls out to a ccall that uses these condition registers to figure out what the result of the branch should be.

However! vex's optimizer certain cases so that it doesn't have to do the ccall, instead inlining the actual computation of just the flag that matters. It can only do this when the comparison and the branch are in the same basic block. Therefore, using num_inst=1 will disable vex's optimization and force ccall usage, and using fastpath will prevent this from actually working.

Basically, don't use fastpath. It's meant for static analysis, not emulation.

**Comment by symike** [comment]
Changing the default analysis mode to symbolic (from fastpath), to ensure that the `DO_CCALLS` option is turned on (unlike in fastpath, where it's one of the options removed), sadly does not solve the issue.

**Comment by rhelmot** [comment]
In that case I can't reproduce. Removing "fastpath" from your example script makes it successfully run to the end.
